libraries/shared/logs/extractor.go

```diff
@@ -0,0 +1,132 @@
+
+type ILogExtractor interface {
+	AddTransformerConfig(config transformer.EventTransformerConfig)
+	ExtractLogs(recheckHeaders constants.TransformerExecution, errs chan error, missingHeadersFound chan bool)
+}
+
+type LogExtractor struct {
+	Addresses                []common.Address
+	CheckedHeadersRepository datastore.CheckedHeadersRepository
+	Fetcher                  fetcher.ILogFetcher
+	LogRepository            datastore.HeaderSyncLogRepository
+	StartingBlock            *int64
+	Syncer                   transactions.ITransactionsSyncer
+	Topics                   []common.Hash
+}
+
+// Add additional logs to extract
+func (extractor *LogExtractor) AddTransformerConfig(config transformer.EventTransformerConfig) {
+	if extractor.StartingBlock == nil {
+		extractor.StartingBlock = &config.StartingBlockNumber
+	} else if earlierStartingBlockNumber(config.StartingBlockNumber, *extractor.StartingBlock) {
+		extractor.StartingBlock = &config.StartingBlockNumber
+	}
+
+	addresses := transformer.HexStringsToAddresses(config.ContractAddresses)
+	extractor.Addresses = append(extractor.Addresses, addresses...)
+	extractor.Topics = append(extractor.Topics, common.HexToHash(config.Topic))
+}
+
+// Fetch and persist watched logs
+func (extractor LogExtractor) ExtractLogs(recheckHeaders constants.TransformerExecution, errs chan error, missingHeadersFound chan bool) {
+	if len(extractor.Addresses) < 1 {
+		logrus.Errorf("error extracting logs: %s", ErrNoWatchedAddresses.Error())
+		errs <- ErrNoWatchedAddresses
+		return
+	}
+
+	missingHeaders, missingHeadersErr := extractor.CheckedHeadersRepository.MissingHeaders(*extractor.StartingBlock, -1, getCheckCount(recheckHeaders))
+	if missingHeadersErr != nil {
+		logrus.Errorf("error fetching missing headers: %s", missingHeadersErr)
+		errs <- missingHeadersErr
+		return
+	}
+
+	if len(missingHeaders) < 1 {
+		missingHeadersFound <- false
+		return
+	}
+
+	for _, header := range missingHeaders {
+		logs, fetchLogsErr := extractor.Fetcher.FetchLogs(extractor.Addresses, extractor.Topics, header)
+		if fetchLogsErr != nil {
+			logError("error fetching logs for header: %s", fetchLogsErr, header)
+			errs <- fetchLogsErr
+			return
+		}
+
+		if len(logs) > 0 {
+			transactionsSyncErr := extractor.Syncer.SyncTransactions(header.Id, logs)
+			if transactionsSyncErr != nil {
+				logError("error syncing transactions: %s", transactionsSyncErr, header)
+				errs <- transactionsSyncErr
+				return
+			}
+
+			createLogsErr := extractor.LogRepository.CreateHeaderSyncLogs(header.Id, logs)
+			if createLogsErr != nil {
+				logError("error persisting logs: %s", createLogsErr, header)
+				errs <- createLogsErr
+				return
+			}
+		}
+
+		markHeaderCheckedErr := extractor.CheckedHeadersRepository.MarkHeaderChecked(header.Id)
+		if markHeaderCheckedErr != nil {
+			logError("error marking header checked: %s", markHeaderCheckedErr, header)
+			errs <- markHeaderCheckedErr
```

Comment:
Just came up with a resilience improvement idea btw:
In general, when we do things like this that _should_ succeed but probably fails due to temporary network issues/problems on another machine (DB box), couldn't we maintain a local work list of failed requests and try to re-send them?

Reply:
Maybe, but that feels like perhaps unnecessary complexity? Like, shouldn't this just get automatically run on a subsequent pass (where it'll presumably succeed)?

Reply:
@rmulhol Oh right, it will because this is outside the `logs > 0` conditional :blush: